pkg/blockbuilder/blockbuilder.go

```diff
@@ -239,3 +238,267 @@ func (b *BlockBuilder) getLagForPartition(ctx context.Context, partition int32)
+		b.cfg.Kafka.Topic: {
+			partition: kgo.NewOffset().At(state.Commit.At),
+		},
+	})
+	defer b.kafkaClient.RemoveConsumePartitions(map[string][]int32{b.cfg.Kafka.Topic: {partition}})
+
+	level.Info(logger).Log("msg", "start consuming", "offset", state.Commit.At)
+
+	var (
+		firstRec  *kgo.Record
+		lastRec   *kgo.Record
+		commitRec *kgo.Record
+	)
+
+consumerLoop:
+	for remaining := state.Lag; remaining > 0; {
+		if err := context.Cause(ctx); err != nil {
+			return partitionState{}, err
+		}
+		fetches := b.kafkaClient.PollFetches(ctx)
+		fetches.EachError(func(_ string, _ int32, err error) {
+			if !errors.Is(err, context.Canceled) {
+				level.Error(logger).Log("msg", "failed to fetch records", "err", err)
+				b.blockBuilderMetrics.fetchErrors.Inc()
+			}
+		})
+
+		numRecs := fetches.NumRecords()
+		remaining -= int64(numRecs)
+
+		recIter := fetches.RecordIter()
+		for !recIter.Done() {
+			rec := recIter.Next()
+
+			if firstRec == nil {
+				firstRec = rec
+			}
+
+			// Stop consuming after we reached the cycleEnd marker.
+			// NOTE: the timestamp of the record is when the record was produced relative to distributor's time.
+			if rec.Timestamp.After(cycleEnd) {
+				break consumerLoop
+			}
+
+			recProcessedBefore := rec.Offset <= state.LastSeenOffset
+			allSamplesProcessed, err := builder.Process(ctx, rec, state.LastBlockEnd.UnixMilli(), blockEnd.UnixMilli(), recProcessedBefore)
+			if err != nil {
+				// All "non-terminal" errors are handled by the TSDBBuilder.
+				return state, fmt.Errorf("process record in partition %d at offset %d: %w", rec.Partition, rec.Offset, err)
+			}
+			if !allSamplesProcessed {
+				if lastRec == nil {
+					// The first record was not fully processed, meaning the record before this is the commit point.
+					// We hand-craft the commitRec from the data in the state to re-commit it. On commit the commit's meta is updated
+					// with the new value of LastSeenOffset. This is so the next cycle handled partially processed record properly.
+					commitRec = &kgo.Record{
+						Topic:       state.Commit.Topic,
+						Partition:   state.Commit.Partition,
+						Offset:      state.Commit.At - 1, // Previous commit's offset-1 means we expect the next cycle to start from the commit's offset.
+						Timestamp:   state.CommitRecTimestamp,
+						LeaderEpoch: state.Commit.LeaderEpoch,
+					}
+				} else if commitRec == nil {
+					// The commit offset should be the last record that was fully processed and not the first record that was not fully processed.
+					commitRec = lastRec
+				}
+			}
+			lastRec = rec
+		}
+	}
+
+	// Nothing was consumed from Kafka at all.
+	if firstRec == nil {
+		level.Info(logger).Log("msg", "no records were consumed")
+		return state, nil
+	}
+
+	// No records were processed for this cycle.
+	if lastRec == nil {
+		level.Info(logger).Log("msg", "nothing to commit due to first record fetched is from next cycle", "first_rec_offset", firstRec.Offset)
```

Comment:
[nit] I would clarify it. I would also log the firstRec timestamp and cycle end.

```suggestion
		level.Info(logger).Log("msg", "nothing to commit due to first record fetched has a timestamp greater than this cycle end", "first_rec_offset", firstRec.Offset)
```